src/main/flight/imu.c

```diff
@@ -215,51 +213,79 @@ static float invSqrt(float x)
+        const float headingMagnitude = vector2Mag(&heading_ef);
+
+        // cross (product) = heading * cogHeading * sin error angle, ie, reflects error angle of quad X axis vs GPS groundspeed
+        // cross value sign depends on the rotation change from input to output vectors by right hand rule
+        // operand order is such that rotation is in direction of zero error
+        // abs value of cross is zero when parallel, max of 1.0 at 90 degree error
+        // cross value determines ez_ef when error is less than 90 degrees
+        // a zero cross value at 180 degree error would prevent error correction
         const float cross = vector2Cross(&heading_ef, &cog_ef);
-        // dot product, 1 * |heading| * cos(angle)
+
+        // dot product = heading * cogHeading * cos(angle)
+        // max when aligned, zero at 90 degrees of vector difference, negative for any error > 90 degrees
         const float dot = vector2Dot(&heading_ef, &cog_ef);
-        // use cross product / sin(angle) when error < 90deg (cos > 0),
-        //   |heading| if error is larger (cos < 0)
-        const float heading_mag = vector2Mag(&heading_ef);
-        float ez_ef = (dot > 0) ? cross : (cross < 0 ? -1.0f : 1.0f) * heading_mag;
-#if THRUST_COG
-        // increase gain for small tilt (just heuristic; sqrt is cheap on F4+)
-        ez_ef /= sqrtf(heading_mag);
-#endif
-        ez_ef *= cogYawGain;          // apply gain parameter
+
+
+        // ez_ef is the error value used to correct the heading via an IIR like smoothing function
+        // for error angles less than 90 degrees (dot > 0), use cross value to compute ez_ef
+        // above 90 degrees error, use heading vector magnitude, but give it sign from cross product
+        // heading vector magnitude is 1.0 when flat, and is attenuated when pitch angle is large
+        // when error exceeds 90 degrees, and pitch angle is high, a transient step may occur in ez_ef
+        float ez_ef = (dot > 0) ? cross : (cross < 0 ? -1.0f : 1.0f) * headingMagnitude;
+
+        // avoid using GPS (reduce ez_ef) during and soon after roll or yaw inputs
+        // because the GPS heading will tend to lag the true heading
+        const float deflectionRPMaxInv = 1.0f - fmaxf(getRcDeflectionAbs(FD_ROLL), getRcDeflectionAbs(FD_YAW));
+        float deflectionRPMaxInvPwr5 = power5(deflectionRPMaxInv);
+        // negative peak detector with 1s time constant 1.0s
+        static float prev = 0.0f;
+
+        const float k = 0.5f * dt; // k = 0.005 at 100Hz, dt in seconds, 2.0s time constant
+        float smoothedStickMovement = prev + k * (deflectionRPMaxInvPwr5 - prev);
+        prev = fminf(smoothedStickMovement, deflectionRPMaxInvPwr5);
+
+        const fpVector2_t thrustVectorEF = {.x = rMat[X][Z], .y = rMat[Y][Z]};
+        // body Z axis (up) - direction of thrust vector - 0-1
+        // magnitude is closer to 1 when thrust is 'forward', closer to 0 when thrust is 'backward'
```

Comment:
zero when level, 1 at +-90 pitch.

Reply:
I was hoping for a value that would be positive when the thrust vector would cause acceleration of the craft in the earth frame, decreasing to zero when there is no contribution (quad is flat to earth frame), and that does not increase above zero when the thrust vector would cause deceleration of the quad in its forward earth frame direction.

It sounds like thrustVectorEF needs a way to stop it rising when thrust is 'backward'.

Reply:
You can project earth-frame Up vector into body coordinates and then on X axis.
IIRC Z axis projection is {.x = rMat[Z][X], .y = rMat[Z][Y], .z = rMat[Z][Z] } and you only want X component, so rMat[Z][X]. But I did not test it

Reply:
My problem is that these abstraction levels are over my head.  I do understand the flight requirements, very well, but matrix maths is something I never grasp properly.

Reply:
This ChatGPT answer is quite useful. I did not check, maybe rows/columns are swapped in BF:

### Understanding the Rotation Matrix

The rotation matrix \(R\) is a square matrix used to rotate vectors in a three-dimensional space. It is composed of elements that directly correspond to the axes of the two coordinate systems involved: the body frame (local frame attached to a moving body, for example, an aircraft or a drone) and the Earth frame (a fixed, global frame, often aligned with the North, East, and Down directions).

In a rotation matrix \(R\), each column represents the unit vectors of the body frame axes expressed in the Earth frame. Conversely, each row represents the unit vectors of the Earth frame axes expressed in the body frame. This duality is key to understanding how the rotation matrix facilitates the transformation between the two frames.

### Practical Implications

1. **Column Interpretation**:
   - **First Column**: Represents the orientation of the body's x-axis in Earth coordinates. Essentially, it tells you how much of the body's x-axis aligns with the Earth's x, y, and z axes.
   - **Second Column**: Represents the orientation of the body's y-axis in Earth coordinates, similarly describing its alignment with the Earth's axes.
   - **Third Column**: Represents the orientation of the body's z-axis in Earth coordinates, again describing its alignment.

2. **Row Interpretation**:
   - **First Row**: Tells you how the Earth's x-axis is represented in the body frame, indicating the projection of the Earth's x-axis onto the body's axes.
   - **Second Row**: Similar to the first, but for the Earth's y-axis.
   - **Third Row**: For the Earth's z-axis.

### Practical Use

- **Vector Transformation**: If you have a vector expressed in the body frame and you want to know its coordinates in the Earth frame, you multiply the rotation matrix \(R\) by this vector. This operation essentially sums up the contributions of each body axis to the Earth frame axes.
- **Understanding Orientation**: By examining the rotation matrix, you can deduce the orientation of the body relative to the Earth. For example, if one of the columns is primarily aligned with one of the Earth's axes, the body is mainly oriented along that Earth axis.
- **Sensor Data Interpretation**: For sensors fixed on a body (like a drone), their readings (acceleration, magnetic field, etc.) are in the body frame. To interpret these readings correctly in a global context (e.g., for navigation), you need to transform them into the Earth frame using the rotation matrix.

Reply:
Thanks.